### ISSUE TITLE
bug: out of memory because model switching not accounting for clip models

### ISSUE THREAD

**Issue by rcarmo** [brycedrennan/imaginAIry#106]
Quick question: how do I switch the model programmatically (like when running this inside JupyterLab?) Do I need to explicitly unload and reload another model? 

Specifying a new model per prompt just runs out of memory, so I assume there ought to be some way to unload the models in Python.

**Comment by brycedrennan** [comment]
It's supposed to unload and load the models as needed but it sounds like that may not be working. That's probably not super well tested.   I wonder if it's related to running in Jupyter and it would work just as a python script.

**Comment by rcarmo** [comment]
I'm restarting the kernel, but every time I try to run a prompt with a second model it fails. The GPUs I'm using only have 8GB, and space seems pretty tight in there via `nvidia-smi`

**Comment by rcarmo** [comment]
There might be more problems. Look what happens when I try using `imagine` twice in the same session with a second model:

```bash
me@gpu:~$ imagine "an engineer coding at a computer desktop, wide angle, intricate detail, artwork by moebius" --width 640 --height 384
🤖🧠 imaginAIry received 1 prompt(s) and will repeat them 1 times to create 1 images.
Generating 🖼  1/1: "an engineer coding at a computer desktop, wide angle, intricate detail, artwork by moebius" 640
x384px seed:423675749 prompt-strength:7.5 steps:15 sampler-type:k_dpmpp_2m
Loading model /home/me/.cache/huggingface/transformers/24bd254e54b30e83bcbc15efae29f0ef55256fd144823a9437f5956e594f6803.dcd6f0dd97c55495efb8393e64f704f3f398d695965876be0339bf96b93e2b4e onto cuda backend...
100%|███████████████████████████████████████████████████████████████████████████████| 15/15 [00:11<00:00,  1.36it/s]
    Image Generated. Timings: conditioning:0.36s sampling:11.04s safety-filter:3.90s total:16.05s
    🖼  [generated] saved to: ./outputs/generated/000003_423675749_kdpmpp2m15_PS7.5_an_engineer_coding_at_a_computer
_desktop,_wide_angle,_intricate_detail,_artwork_by_moebius_[generated].jpg
me@gpu:~$ imagine "an engineer coding at a computer desktop, wide angle, intricate detail, artwork by moebius" --width 640 --height 384 --model SD-1.5
🤖🧠 imaginAIry received 1 prompt(s) and will repeat them 1 times to create 1 images.
Generating 🖼  1/1: "an engineer coding at a computer desktop, wide angle, intricate detail, artwork by moebius" 640
x384px seed:602625425 prompt-strength:7.5 steps:15 sampler-type:k_dpmpp_2m
Loading model /home/me/.cache/huggingface/transformers/4c1a32af58eeaff9f36410f7ca27e51a8856185c3f05d5b930975a1397914f10.98fc1312797017a8bac6993df565908fd18f09319b40d9bd35457dfa1459ecf0 onto cuda backend...
/usr/lib/python3/dist-packages/apport/report.py:13: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
  import fnmatch, glob, traceback, errno, sys, atexit, imp, stat
Traceback (most recent call last):
  File "/usr/local/bin/imagine", line 8, in <module>
    sys.exit(imagine_cmd())
  File "/usr/lib/python3/dist-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/cmds.py", line 238, in imagine_cmd
    imagine_image_files(
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/api.py", line 74, in imagine_image_files
    for result in imagine(
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/api.py", line 151, in imagine
    neutral_conditioning = model.get_learned_conditioning(
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/modules/diffusion/ddpm.py", line 797, in get_learned_conditioning
    c = self.cond_stage_model.encode(c)
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/modules/clip_embedders.py", line 49, in encode
    return self(text)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/imaginairy/modules/clip_embedders.py", line 43, in forward
    outputs = self.transformer(input_ids=tokens)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 722, in forward
    return self.text_model(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 643, in forward
    encoder_outputs = self.encoder(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 574, in forward
    layer_outputs = encoder_layer(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 317, in forward
    hidden_states, attn_weights = self.self_attn(
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/transformers/models/clip/modeling_clip.py", line 209, in forward
    query_states = self.q_proj(hidden_states) * self.scale
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1190, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/linear.py", line 114, in forward
    return F.linear(input, self.weight, self.bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling `cublasCreate(handle)`
```

**Comment by brycedrennan** [comment]
Yeah that's an out of memory error.  I likely cause is that there are different clip models, which I am not currently facilitating unloads for.  The new clip model is bigger.   So you should be able to switch between 2.0 models in the same session or 1.0 models in the same session - but not both.

**Comment by brycedrennan** [comment]
I don't think I want to invest a lot of time in addressing this so I'll probably just add a warning about this limitation.

**Comment by rcarmo** [comment]
A way to unload everything and reset the GPU would probably be enough…On 26 Nov 2022, at 19:36, Bryce Drennan ***@***.***> wrote:﻿
I don't think I want to invest a lot of time in addressing this so I'll probably just add a warning about this limitation.

**Comment by brycedrennan** [comment]
I ran into this as well on my computer which means it's gonna be a priority now :-)

**Comment by ronitrocket** [comment]
I'm encountering this issue as well on google colab. I've tried completely deleting the runtime and trying again but still cannot get the image to finish generating before the RAM is fully allocated and it crashes.

**Comment by brycedrennan** [comment]
Fixed in 7.1.1
